producer-c/producer-cloudwatch-integ/KvsProducerSampleCloudwatch.cpp

```diff
@@ -180,26 +210,53 @@ INT32 main(INT32 argc, CHAR *argv[])
             frame.presentationTs = frame.decodingTs;
             frameIndex++;
         }
+        CHK_LOG_ERR(retStatus);
+
+        if (frame.frameData != NULL) {
+            SAFE_MEMFREE(frame.frameData);
+        }
 
+        freeDeviceInfo(&pDeviceInfo);
+        freeStreamInfoProvider(&pStreamInfo);
+        freeKinesisVideoStream(&streamHandle);
+        freeKinesisVideoClient(&clientHandle);
+        freeCallbacksProvider(&pClientCallbacks); // This will also take care of freeing canaryStreamCallbacks
+        DLOGI("Memory allocated:%llu bytes", getInstrumentedTotalAllocationSize());
+        RESET_INSTRUMENTED_ALLOCATORS();
+        DLOGI("CleanUp Done");
+        cleanUpDone = TRUE;
+        if(!fileLoggingEnabled) {
+            // This is necessary to ensure that we do not lose the last set of logs
+            canaryStreamSendLogSync(&cloudwatchLogsObject);
+        }
     }
 CleanUp:
-
     Aws::ShutdownAPI(options);
+    CHK_LOG_ERR(retStatus);
 
-    if (STATUS_FAILED(retStatus)) {
-        DLOGE("Failed with status 0x%08x\n", retStatus);
-    }
+    // canaryStreamSendLogSync() will lead to segfault outside the block scope
+    // https://docs.aws.amazon.com/sdk-for-cpp/v1/developer-guide/basic-use.html
+    // The clean up related logs also need to be captured in cloudwatch logs. This flag
+    // will cater to two scenarios, one, if any of the commands fail, this clean up is invoked
+    // Second, it will cater to scenarios where a SIG is sent to exit the while loop above in
+    // which case the clean up related logs will be captured as well.
+    if(!cleanUpDone) {
+        if (STATUS_FAILED(retStatus)) {
```

Comment:
You are already doing CHK_LOG_ERR just above

Reply:
As per comment, I am doing these set of operations twice to allow logging to cloudwatch if I receive a signal. And the second time is not executed if I execute the first time.